### ISSUE TITLE
Consider simplifying Packet Number Encryption

### ISSUE THREAD

**Issue by DavidSchinazi** [quicwg/base-drafts#1575]
Currently PNE encrypts the variable-length packet number. There are at least two ways to implement decryption:

1. decrypt the first byte in-place, compute the PN length, decrypt the remaining bytes in-place
2. decrypt the max PN length (4 bytes) into a separate buffer, compute the PN length, copy that amount of bytes back

This has the downside of worse performance, or can cause timing attacks.

It would be simpler to always encrypt the max length - that way we can always decrypt 4 bytes in place.

Apologies if this has been discussed before, I wasn't able to find it.

**Comment by mikkelfj** [comment]
If you send very short messages such as PING you risk overlapping with the AEAD tag and then it starts to get interesting figuring out how to decrypt the packet number.

**Comment by DavidSchinazi** [comment]
I'm not sure I understand, can you elaborate? You just decrypt four bytes - having that flow into the tag is not a problem

**Comment by martinthomson** [comment]
I think that this is purely editorial: the draft doesn't actually describe a process for decryption, just points at considerations.  Are you suggesting that we recommend a specific method?  I think that any of these options are fine.

**Comment by DavidSchinazi** [comment]
Sorry, I meant that we could change the amount we encrypt - always encrypt four bytes instead of just the PN.

But thinking about this some more I understand @mikkelfj 's comment. If we encrypt the sample we'll have a hard time decrypting... Would it be reasonable to reduce the size of the sample for very short packets?

**Comment by martinthomson** [comment]
@DavidSchinazi , I see.  We shouldn't really reduce the sample size.  That is, if you value the strength of the encryption.  Either way, it shouldn't be necessary.

**Comment by DavidSchinazi** [comment]
I agree that this isn't necessary - I'm just wondering if this proposal could make implementations simpler and safer.

Proposal: always encrypt four bytes [pn_offset, pn_offset + 3] and always have the sample start at pn_offset + 4. If the sample were to overflow the payload + tag (which only happens if you use 128-bit AES and have pn_len + payload_len < 4), you just pad the sample with zeroes. In the worst case scenario the sample contains two bytes of zero but you still have a 128bit secret key. Is that unacceptable?

**Comment by martinthomson** [comment]
I wouldn't say unacceptable, but it does lower the birthday bound.

**Comment by mnot** [comment]
Discussed in NYC; @ekr to follow up with @DavidSchinazi within two weeks.

**Comment by kazuho** [comment]
I prefer the current approach.

IIUC, the proposal is to have a conditional branch that tests if the sum of PN length and payload is below 20 octets, then create a zero-padded copy for sampling.

I do not think that is simpler than the current approach that uses a conditional branch to determine the sampling offset.

**Comment by DavidSchinazi** [comment]
I had a conversation with @ekr about this yesterday, and I'm wondering if the following proposal might get us the best of both worlds:
1) Require that QUIC packets MUST verify `length(packet number) + length(payload) >= 4`, by padding the payload if needed.
2) Always have PNE encrypt 4 bytes from the start of the packet number, and use the following 16 bytes as nonce.

In practice the padding overhead here is very low as it only impacts a small number of frame types when sent without an ACK frame. (Worse case scenario is a lonely PING frame with packet number below 63, and that case suffers 2 bytes of padding out of a 20+8+1+1+1=31 byte IPv4 packet)

It also has the nice property of obfuscating the packet-number length, as an observer can no longer tell when the packet number is below 63.

In terms of implementation, this is simplest for decryption, and the padding can be added right before encrypting the payload as that code needs to have access to the packet number as nonce anyway.

What are people's thoughts on this?

**Comment by martinthomson** [comment]
That sounds good.  We can stipulate that a packet MUST always contain at least 3 octets of frames.  Adding a couple of padding octets is probably a good idea.

Note that we're probably changing the encoding of packet numbers (again, yeah, sorry), so the precise set of 

Not sure whether you mean that you always XOR 4 octets (which means XOR over the encrypted payload if the packet number is <4), or whether you only XOR the octets of the packet number.  With the proposed move of the packet number length to the first octet, it's probably better to only XOR the octets of the packet number.  That way you don't touch the payload octets other than to sample them.

**Comment by DavidSchinazi** [comment]
I think requiring 3 octets of frames is more restrictive that we need to be. I would prefer saying that QUIC packets MUST verify `length(packet number) + length(payload) >= 4`, and if implementors want to do that by always ensuring 3 bytes of frames that's fine.

I meant that we XOR over the payload if the packet number length is < 4. When we're encrypting the PN, I'm not sure the distinction between PN and payload is relevant - there's nothing wrong with changing the start of the payload in my mind. And it simplifies the implementation.

But we can put this on hold until we've landed on what the PN encoding looks like.

**Comment by kazuho** [comment]
> I think requiring 3 octets of frames is more restrictive that we need to be. I would prefer saying that QUIC packets MUST verify `length(packet number) + length(payload) >= 4`, and if implementors want to do that by always ensuring 3 bytes of frames that's fine.

This sounds like a neat approach.

> I meant that we XOR over the payload if the packet number length is < 4.

I am not sure if I like the idea. Because it might be an additional complexity to the hardware decoders. In the current design, payload is decrypted only once by an AEAD cipher. This change would mean that it can be decrypted twice, first by a CTR cipher and then by an AEAD cipher.

As @martinthomson points out, we discussed in New York about moving the PN length bits to the first octet. Assuming that we adopt that change, I think there would be no reason to prefer always CTR-decrypting 4 octets. The different is just having a different order between picking the specified length and applying CTR.

**Comment by DavidSchinazi** [comment]
Could someone please point me at the issue or PR for the PN length bits change proposed in New York, if it exists?

**Comment by kazuho** [comment]
Minutes with a link to the slides is available at https://github.com/quicwg/wg-materials/blob/master/interim-18-09/minutes.md#first-octet---martin-thomson.
